### ISSUE TITLE
zstdcli.c memory checks, invalid read when using *just* "-" , leak from filenameTable when run with no args or input.

### ISSUE THREAD

**Issue by thatsafunnyname** [Cyan4973/zstd#153]
Hello and thank you very much for Zstandard,
 
While running some tests using zstd I had some memory checks alert:
   * When zstd is run without args or stdin input, filenameTable is not free'ed.
   * When zstd is passed *just* "-" then there is an invalid read when strcmp is called on outFileName.

Details below.
 
Cheers.
 
    > cd programs
    > make CFLAGS=-g
    > valgrind --leak-check=yes ./zstd
    
    HEAP SUMMARY:
    in use at exit: 8 bytes in 1 blocks
    total heap usage: 1 allocs, 0 frees, 8 bytes allocated
    8 bytes in 1 blocks are definitely lost in loss record 1 of 1
    at 0x4A06A2E: malloc (vg_replace_malloc.c:270)
    by 0x4587E4: main (zstdcli.c:182)
    
    LEAK SUMMARY:
    definitely lost: 8 bytes in 1 blocks
    indirectly lost: 0 bytes in 0 blocks
    possibly lost: 0 bytes in 0 blocks
    still reachable: 0 bytes in 0 blocks
    suppressed: 0 bytes in 0 blocks

Line 182 is

    const char** filenameTable = (const char**)malloc(argCount * sizeof(const char*));   /* argCount >= 1 */

the free in the `_end` goto is never called.

    > echo -n "Xyzzy" | valgrind --leak-check=full --show-reachable=yes --leak-check=yes ./zstd - | base64
    Memcheck, a memory error detector
    Copyright (C) 2002-2012, and GNU GPL'd, by Julian Seward et al.
    Using Valgrind-3.8.1 and LibVEX; rerun with -h for copyright info
    Command: ./zstd -
    
    Invalid read of size 1
    at 0x3E04328350: __strcmp_sse42 (in /lib64/libc-2.12.so)
    by 0x45920A: main (zstdcli.c:400)
    Address 0x0 is not stack'd, malloc'd or (recently) free'd
    
    Process terminating with default action of signal 11 (SIGSEGV): dumping core
    Access not within mapped region at address 0x0
    at 0x3E04328350: __strcmp_sse42 (in /lib64/libc-2.12.so)
    by 0x45920A: main (zstdcli.c:400)
    If you believe this happened as a result of a stack
    overflow in your program's main thread (unlikely but
    possible), you can try to increase the size of the
    main thread stack using the --main-stacksize= flag.
    The main thread stack size used in this run was 10485760.
    
    HEAP SUMMARY:
    in use at exit: 16 bytes in 1 blocks
    total heap usage: 1 allocs, 0 frees, 16 bytes allocated
    
    16 bytes in 1 blocks are still reachable in loss record 1 of 1
    at 0x4A06A2E: malloc (vg_replace_malloc.c:270)
    by 0x4587E4: main (zstdcli.c:182)
    
    LEAK SUMMARY:
    definitely lost: 0 bytes in 0 blocks
    indirectly lost: 0 bytes in 0 blocks
    possibly lost: 0 bytes in 0 blocks
    still reachable: 16 bytes in 1 blocks
    suppressed: 0 bytes in 0 blocks
 
Line 400 is
 
    if (!strcmp(filenameTable[0], stdinmark) && !strcmp(outFileName,stdoutmark) && (displayLevel==2)) displayLevel=1;
 
I added a check for outFileName being true:
 
    if (!strcmp(filenameTable[0], stdinmark) && outFileName && !strcmp(outFileName,stdoutmark) && (displayLevel==2)) displayLevel=1;

**Comment by Cyan4973** [comment]
Thanks for notification @thatsafunnyname !

There are 2 kind of issues mentioned : 

- Some memory not freed at the end : 
Correct. Although, imho, it's probably not a big issue. After all, the program exits at this point : it leaves `main()`. So, by definition, all program memory is going to be released.
So it's not the same as a _memory leak_, which is a more important problem to track. Here, memory is allocated only once, at the beginning of `main()`, and disappear at the end.
The problem is, there are multiple exit points. Making sure this table is explicitly released from every possible exit point will just make the program more complex to read, without really providing any worthwhile protection.

- `strcmp()` on `NULL` : indeed , thanks for the find and precise reproduction case !
Your proposed protection works fine.
Alternatively, initializing `outFileName` with a zero-length string `= "";` rather than `NULL` could do the trick. But it would be necessary to first check if `outFileName == NULL` is used as a kind of signal in the program.

**Comment by Cyan4973** [comment]
Latest update of "dev060" branch should fix this issue.

I've used your suggested `outFileName` test to avoid `strcmp()` on `NULL`.
I've also added a macro `CLEAN_RETURN` which encapsulates all `return` exit point of `main()` to force them through `_end:` where dynamically allocated structures are freed.

The test suite has also been update, `make valgrindTest` now has new tests checking for the 2 bugs you reported.

Thanks and best regards

**Comment by thatsafunnyname** [comment]
Great, thank you.

Just to provide the background - the build systems I am using run tests with memory checks enabled (valgrind and others), zstd is run as part of a test, the change means I do not have to add suppressions for various memory check tools and it saves other developers from spending time looking at diagnostics reported by the tools (if the suppression config is not present or does not (yet) exist for future memory check tools used).
 
I pulled your update to the "dev060" branch and can confirm that the memory checks no longer alert.
 
I noticed the variable dynNameSpace in programs/zstdcli.c::main looks to be unused, I was wondering if it could be removed?  I can open a new issue for this if you wish.
 
I thought you might like to know Zstandard compiles and runs on OpenVMS running on Itanium, but something is not quite right at the moment as I see a very poor compression ratio in a simple test, if you can help by offering some suggestions then I can open a new issue for this.

Thanks again.
